src/google/protobuf/descriptor.cc

```diff
@@ -5480,6 +5489,75 @@ void DescriptorBuilder::BuildMessage(const DescriptorProto& proto,
   }
 }
 
+std::string GetJSONName(const FieldDescriptorProto& field, bool use_custom, bool* was_custom) {
+  if (use_custom && field.has_json_name()) {
+    *was_custom = true;
+    return field.json_name();
+  }
+  *was_custom = false;
+  return ToJsonName(field.name());
+}
+
+void DescriptorBuilder::CheckFieldJSONNameUniqueness(
+    const DescriptorProto& proto, const Descriptor* result) {
+  bool is_proto2 = result->file()->syntax() == FileDescriptor::SYNTAX_PROTO2;
+  std::string message_name = result->full_name();
+  // two passes: one looking only at default JSON names, and one that considers custom JSON names
+  CheckFieldJSONNameUniqueness(message_name, proto, is_proto2, false);
+  CheckFieldJSONNameUniqueness(message_name, proto, is_proto2, true);
+}
+
+struct JSONNameDetails {
+  const FieldDescriptorProto* field;
+  std::string orig_name;
+  bool is_custom;
+};
+
+void DescriptorBuilder::CheckFieldJSONNameUniqueness(
+    std::string message_name,const DescriptorProto& message, bool is_proto2, bool use_custom_names) {
```

Comment:
You may need to run clang-format here.

Reply:
Done.

I applied clang-format changes for all of my new code with one exception: it wanted to move `&` and `*` to be adjacent to the var/param instead of the type (e.g. `DescriptorProto& message` -> `DescriptorProto &message`). But the rest of this file uses the convention as seen here. I can change all of these, too, but figured it might be better (for consistency) to leave them looking like the rest of the file. Let me know what you think.

Reply:
Yes, please leave them like the rest of the file. It is possible that there are some issues with your clang-format config.